subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/modulecache/ModuleMetaDataCache.java

```diff
@@ -33,7 +34,7 @@
     interface CachedMetaData {
         ResolvedModuleVersion getModuleVersion();
 
-        ModuleComponentResolveMetadata getMetaData();
+        MutableModuleComponentResolveMetadata getMetaData();
```

Comment:
As above, it appears to be internal to the cache and only on the interface side, and prevents immutable -> mutable conversions, which is where the expense is.

Specifically, since I'm not sure I mentioned it:

 `org.gradle.internal.component.external.model.AbstractModuleComponentResolveMetadata#AbstractModuleComponentResolveMetadata(org.gradle.internal.component.external.model.MutableModuleComponentResolveMetadata, org.gradle.internal.component.model.ModuleSource)` due to `org.gradle.internal.component.external.model.AbstractModuleComponentResolveMetadata#populateConfigurationsFromDescriptor`.